content/en/blog/2023/exponential-histograms.md

```diff
@@ -0,0 +1,196 @@
+
+This brings me to the first most important point of choosing a scale, _data
+contrast_. Data contrast is how you describe the difference in scale between the
+smallest possible value x and the largest possible value y in your dataset and
+is calculated as the constant multiple c such that `y = c * x`. For example, if
+your data is between 1 and 1000 milliseconds, your data contrast is 1000. If
+your data is between 1 kilobyte and 1 terabyte, your data contrast is
+1,000,000,000. Data contrast, scale, and the number of buckets are all
+interlinked such that if you have 2, you can calculate the third.
+
+Fortunately, if you are using OpenTelemetry, scale choice is largely done for
+you. In OpenTelemetry, you configure a maximum scale (default 20) and a maximum
+size (default 160), or number of buckets, in the histogram. The histogram is
+initially assumed have the maximum scale. As additional data points are added,
+the histogram will rescale itself down such that the data points always fit
+within your maximum number of buckets. The default of 160 buckets was chosen by
+the OpenTelemetry authors to be able to cover typical web requests between 1ms
+and 10s with less than 5% relative error. If your data has less contrast, your
+error will be even less.
+
+# Negative or zero values
+
+For the bulk of this post we have ignored zero and negative values, but negative
+buckets work much the same way, growing larger as the buckets get further from
+zero. All of the math and explanation above applies in the same way to negative
+values, but they should be substituted for their absolute values, and upper
+bounds for buckets are lower bounds (or upper absolute value bounds). Zero
+values, or values with an absolute value less than a configurable threshold, go
+into a special zero bucket. When merging histograms with differing zero
+thresholds, the larger threshold is taken and any buckets with absolute value
+upper bounds within the zero threshold are added to the zero bucket and
+discarded.
+
+# OpenTelemetry and Prometheus
+
+Compatibility between OpenTelemetry and Prometheus is probably a topic large
+enough for its own post. For now I will just say that for all practical
+purposes, OpenTelemetry exponential histograms are 1:1 compatible with
+Prometheus native histograms. Scale calculations, bucket boundaries, error
+rates, zero buckets, etc are all the same. For more information, I recommend you
+watch this talk given by Ruslan Vovalov and Ganesh Vernekar: [Using
+OpenTelemetry’s Exponential Histograms in Prometheus][]
+
+_A version of this article was [originally posted][] to the author's blog._
+
+[originally posted]: {{% param canonical_url %}}
+
+[Why Histograms?]: {{% relref "why-histograms" %}}
+
+[Histograms vs Summaries]: {{% relref "histograms-vs-summaries" %}}
+
+[Using OpenTelemetry’s Exponential Histograms in Prometheus]:
+  https://www.youtube.com/watch?v=W2_TpDcess8
```

Comment:
```suggestion
[Why Histograms?]: ../why-histograms/
[Histograms vs Summaries]: ../histograms-vs-summaries/
[Using OpenTelemetry’s Exponential Histograms in Prometheus]:
  https://www.youtube.com/watch?v=W2_TpDcess8

[originally posted]: {{% param canonical_url %}}
```

Reply:
When I did this before the formatter got mad. It tries to combine all 3 link definition lines into a single paragraph which breaks the markdown

Reply:
This is what `npm run format` does to the markdown in your suggestion:

```
[Why Histograms?]: {{% relref "why-histograms" %}} [Histograms vs Summaries]:
{{% relref "histograms-vs-summaries" %}} [Using OpenTelemetry’s Exponential
Histograms in Prometheus]: https://www.youtube.com/watch?v=W2_TpDcess8

[originally posted]: {{% param canonical_url %}}
```

Reply:
Found a way to make it nicer using prettier ignore directives